pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchMessageContainerImpl.java

```diff
@@ -44,16 +44,17 @@
  * [(k1, v1), (k2, v1), (k3, v1), (k1, v2), (k2, v2), (k3, v2), (k1, v3), (k2, v3), (k3, v3)]
  */
 public class RawBatchMessageContainerImpl extends BatchMessageContainerImpl {
-    MessageCrypto msgCrypto;
-    Set<String> encryptionKeys;
-    CryptoKeyReader cryptoKeyReader;
+    private MessageCrypto<MessageMetadata, MessageMetadata> msgCrypto;
+    private Set<String> encryptionKeys;
+    private CryptoKeyReader cryptoKeyReader;
+    private MessageIdImpl lastAddedMessageId;
 
-    public RawBatchMessageContainerImpl(int maxNumMessagesInBatch, int maxBytesInBatch) {
+    public RawBatchMessageContainerImpl() {
         super();
         this.compressionType = CompressionType.NONE;
         this.compressor = new CompressionCodecNone();
-        this.maxNumMessagesInBatch = maxNumMessagesInBatch;
-        this.maxBytesInBatch = maxBytesInBatch;
+        this.maxNumMessagesInBatch = Integer.MAX_VALUE;
+        this.maxBytesInBatch = Integer.MAX_VALUE;
```

Comment:
It seems that these two fields only affect `haveEnoughSpace` and `isFull`. If your intention is to customize the behavior of these two methods, I think it's better to just override them. Since `hasEnoughSpace` is already overridden, you can override `isFull` with a trivial implementation that returns false.

Reply:
The `isBatchFull` method will be used by `org.apache.pulsar.client.impl.BatchMessageContainerImpl#add`, if we override it, we need to override the `add` method as well.

Reply:
> if we override it, we need to override the add method as well.

Sorry I didn't get it. I mean, if you made the following changes:

```diff
diff --git a/pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchMessageContainerImpl.java b/pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchMessageContainerImpl.java
index a40ab35dab..67df65279b 100644
--- a/pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchMessageContainerImpl.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawBatchMessageContainerImpl.java
@@ -54,8 +54,6 @@ public class RawBatchMessageContainerImpl extends BatchMessageContainerImpl {
         super();
         this.compressionType = CompressionType.NONE;
         this.compressor = new CompressionCodecNone();
-        this.maxNumMessagesInBatch = Integer.MAX_VALUE;
-        this.maxBytesInBatch = Integer.MAX_VALUE;
     }
     private ByteBuf encrypt(ByteBuf compressedPayload) {
         if (msgCrypto == null) {
@@ -98,6 +96,11 @@ public class RawBatchMessageContainerImpl extends BatchMessageContainerImpl {
         return super.add(msg, callback);
     }
 
+    @Override
+    protected boolean isBatchFull() {
+        return false;
+    }
+
     @Override
     public boolean haveEnoughSpace(MessageImpl<?> msg) {
         if (lastAddedMessageId == null) {
```

what else should you make in addition?

Reply:
You can see we called `super.add(msg, callback);`, in this method, it will use the `isBatchFull` method. Override is not working for the parent class.

```
    @Override
    public boolean add(MessageImpl<?> msg, SendCallback callback) {
        this.lastAddedMessageId = (MessageIdAdv) msg.getMessageId();
        return super.add(msg, callback);
    }
```

Reply:
> Override is not working for the parent class.

I think it works and I just verified it with the following demo.

```java
class Base {

    public boolean f() {
        return g();
    }

    protected boolean g() {
        return false;
    }
}

class Derived extends Base {

    @Override
    public boolean f() {
        return super.f();
    }

    @Override
    public boolean g() {
        return true;
    }
}

public class Main {

    public static void main(String[] args) {
        System.out.println(new Base().f());
        System.out.println(new Derived().f());
    }
}
```

Output:

```
false
true
```

Reply:
Oh, yes, you are right.